### PR TITLE
[MRG] Make concatenate_epochs work with empty Epochs objects

### LINKED ISSUE #9533: Concatenation of epochs raises non-intuitive error if one epoch is empty
#### Describe the bug

When concatenating a list of epochs, when one of the epochs contains zero events, a ValueError with few information is raised.

#### Steps to reproduce

```python
import mne
import numpy as np

n_epochs, n_channels, n_timepoints = 10, 4, 21
ch_names = [f"ch_{i}" for i in range(n_channels)]
fake_data_1 = np.random.random((n_epochs, n_channels, n_timepoints))
fake_data_2 = np.random.random((n_epochs, n_channels, n_timepoints))

info = mne.create_info(ch_names=ch_names, sfreq=40)
epo_1 = mne.EpochsArray(fake_data_1, info)
epo_2 = mne.EpochsArray(fake_data_2, info)

# works
conc_epos = mne.concatenate_epochs([epo_1, epo_2])
# remove all data from epo_2
epo_2.drop(range(10))
# Get cryptic ValueError: 
# zero-size array to reduction operation maximum which has no identity
not_conc_epos = mne.concatenate_epochs([epo_1, epo_2])
```

#### Expected results

Either get a concatenated epoch object that contains all events from `epo_1` with a warning that some epos were empty, or get a ValueError that better explains what's happening.

#### Actual results

MNE raises the following error:

```
Traceback (most recent call last):
  File "<input>", line 4, in <module>
  File "/home/jan/dev/python/spot/spot_venv/lib/python3.8/site-packages/mne/epochs.py", line 3493, in concatenate_epochs
    return _finish_concat(*_concatenate_epochs(epochs_list,
  File "/home/jan/dev/python/spot/spot_venv/lib/python3.8/site-packages/mne/epochs.py", line 3419, in _concatenate_epochs
    events_offset = _update_offset(events_offset, epochs.events, shift)
  File "/home/jan/dev/python/spot/spot_venv/lib/python3.8/site-packages/mne/epochs.py", line 3360, in _update_offset
    offset = np.int64(offset) + np.max(events[:, 0]) + shift
  File "<__array_function__ internals>", line 5, in amax
  File "/home/jan/dev/python/spot/spot_venv/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 2705, in amax
    return _wrapreduction(a, np.maximum, 'max', axis, None, out,
  File "/home/jan/dev/python/spot/spot_venv/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 87, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
ValueError: zero-size array to reduction operation maximum which has no identity
```

My current workaround is to manually filter out empty epochs beforehand.

#### Additional information

Platform:      Linux-5.8.0-59-generic-x86_64-with-glibc2.29
Python:        3.8.10 (default, Jun  2 2021, 10:49:15)  [GCC 9.4.0]
Executable:    /home/jan/dev/python/spot/spot_venv/bin/python
CPU:           x86_64: 8 cores
Memory:        Unavailable (requires "psutil" package)
mne:           0.23.0
numpy:         1.19.5 {blas=openblas, lapack=openblas}
scipy:         1.5.4
matplotlib:    3.3.4 {backend=module://backend_interagg}
sklearn:       0.23.2
numba:         Not found
nibabel:       3.2.1
nilearn:       Not found
dipy:          Not found
cupy:          Not found
pandas:        1.1.5
mayavi:        Not found
pyvista:       Not found
vtk:           Not found

### PR DESCRIPTION
**Pull request by wmvanvliet**
There was an edge case when concatenating Epochs and one (or more) of the `Epochs` objects was empty. See #9533.

fixes #9533

### CODE REVIEW

**Comment by wmvanvliet** [comment]
This PR also redoes the logic of the `_update_offset` function which I found very confusing. Do you agree with this new implementation @larsoner?

File: mne/epochs.py [hunk 1]
```diff
@@ -3367,17 +3356,27 @@ def _concatenate_epochs(epochs_list, with_data=True, add_offset=True, *,
         evs = epochs.events.copy()
         # add offset
         if add_offset:
-            evs[:, 0] += events_offset
-        # Update offset for the next iteration.
-        events_offset = _update_offset(events_offset, epochs.events, shift)
+            if len(evs) == 0:
+                warn('One of the Epochs objects to concatenate was empty.')
+            else:
```
Comment (larsoner) [inline mne/epochs.py:3361]: We probably always want to warn, not just when `add_offset == True`, so probably this is better structured as:
```
if len(evs) == 0:
    warn(...)
elif add_offset:
    ...
```

**Comment by larsoner** [comment]
... and agreed the old logic didn't make much sense. No reason to track the offset when it's not being added!

**Review by larsoner — APPROVED** [review]
Can you add a `latest.inc` update in the BUG section? Then let's merge!

**Comment by larsoner** [comment]
... except CIs appear unhappy:
```
mne/tests/test_epochs.py:2706: in test_concatenate_epochs
    concatenate_epochs([epochs, epochs2])
<decorator-gen-237>:24: in concatenate_epochs
    ???
mne/epochs.py:3451: in concatenate_epochs
    return _finish_concat(*_concatenate_epochs(epochs_list,
mne/epochs.py:3364: in _concatenate_epochs
    events_offset += int(np.max(evs[:, 0])) + shift
<__array_function__ internals>:5: in amax
    ???
/usr/share/miniconda/envs/mne/lib/python3.9/site-packages/numpy/core/fromnumeric.py:2754: in amax
    return _wrapreduction(a, np.maximum, 'max', axis, None, out,
/usr/share/miniconda/envs/mne/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
E   ValueError: zero-size array to reduction operation maximum which has no identity
```

**Review by larsoner — CHANGES_REQUESTED** [review]
(Just changing to "request changes" until CIs and latest.inc are taken care of)

**Comment by wmvanvliet** [comment]
I was being stubborn and broke the logic. Now I just took the approach you suggested and everything is working again :)

**Comment by larsoner** [comment]
Thanks @wmvanvliet !

**Pull request merged**